affinity/cpp-23/d1437r0_new.md

```diff
@@ -0,0 +1,271 @@
+
+# Background
+
+Computer systems are no longer homogeneous platforms. Every system from desktop work stations to high-performance supercomputers, and from mobile devices to purpose-built embedded SoCs, every system has some form of co-processor along side the traditional multi-core CPU, and often more than one. Furthermore the architecture these co-processors range from many-core CPUs, GPUs, FPGAs and DSPs to specifically designed vision and machine learning processors. In larger supercomputer systems there are thousands of these processors, connected physically or via network adapters in some configuration of nodes made up of these processors.
+
+The way these processors access memory is also far from homogeneous, there are strict hierarchical memory structures of different address spaces, shared virtual address space with different levels of consistency, cache coherency and atomic operations, NUMA memory regions, persistent memory, HBM memory, and all of there are many different ways in which these memory regions can be configured around the processors.
+
+In order to program these new systems and the architectures that inhabit them, it's vital that applications are capable of understating both what architectures are available and the properties of those architectures; their observable behaviors, capabilities and limitations. However C++ currently has virtually no way to achieve this in the current standard, so developers have to rely entirely on third party and operating system libraries.
+
+# Goals: what this paper is, and what it isn't
+
+This paper seeks to define, within C++, a facility for discovering execution resources available to a system that are capable of executing work, and for querying their properties.
+
+However, it is not the goal of this proposal to introduce support in the C++ language or the standard library for all of the various heterogeneous architectures available today. The authors of this paper recognize that this is unrealistic as it would require significant changes to the C++ machine model and would be extremely volatile to future developments in architecture and system design.
+
+Instead it seeks to define a single unified and stable layer within the C++ standard library, on which applications, libraries and programming models such as standard thread pools, SYCL or Kokkos can be built, and which can be supported, by hardware vendors via standards such as HMM, and extended when necessary.
+
+This layer will not be characterized in terms of specific hardware such as CPUs, GPUs and FPGAs as these are high-level concepts that are not fixed in time and have no foundation in the C++ machine model. It will instead define a number of abstract properties, which reflect an architecture's properties.
```

Comment:
Yes, I think this could be worded better, effectively the point I am trying to get across here is that we don't want to tie the C++ standard to concepts like CPU and GPU, as these concepts are too broad and potentially subject to change, does this make more sense?

Reply:
I think I get what you mean.  Perhaps the last sentence could say: "It will instead define abstract properties of computer architectures that are not tied to specific hardware."